packit_service/models.py

```diff
@@ -2329,6 +2329,33 @@ def add_scan(self, task_id: int) -> "OSHScanModel":
             session.add(scan)
             return scan
 
+    @contextmanager
+    def add_scan_transaction(self, cls) -> Generator["OSHScanModel"]:
+        """
+        Context manager that creates a ScanModel upon entering the context,
+        provides a corresponding instance of `ScanModel` to be updated within the context
+        and commits the changes upon exiting the context, all within a single transaction.
+
+        This locking mechanism is working on the assumption that just a single scan model
+        for build can exist and that we can commit the transaction after having collected
+        inside the context manager the task_id.
```

Comment:
Is this supposed to roll back if the scan already exists, raising an exception? I would expect a check for the existence of a scan.

Reply:
I would expect the context manager to behave somewhat like this, yielding a newly created scan model for updating or `None` if it already exists (the scan has been submitted):

```python
        session = singleton_session or Session()
        session.begin()

        submit_scan = False

        try:
            if not self.scan:
                submit_scan = True
                scan = OSHScanModel()
                scan.copr_build_target = self
                session.add(scan)
        except Exception as ex:
            logger.warning(f"Exception while working with database: {ex!r}")
            session.rollback()
            raise

        try:
            yield scan if submit_scan else None
        except Exception:
            session.rollback()
            raise

        try:
            if submit_scan:
                session.add(scan)
                session.commit()
        except Exception as ex:
            logger.warning(f"Exception while working with database: {ex!r}")
            session.rollback()
            raise
```

Reply:
If we doesn't commit before yielding the scan, we might be doing a "dirty read" and thus we may submit multiple scans. So I don't think I can rely upon a read. From my point of view this works only if we update the db before running the scan... The only thing I see that can ensure us that we will not submit multiple scans is using the db constraints that will raise an exception when updating the db.

As I said I don't like this solution, is tightly coupled with the db design, instead of being just a generic locking mechanism that we could use in a more explicit way.

I would prefer be able to make atomic just the two lines of code I need, but since we are doing it in this way I agree to be consistent:
```
with some_lock:
    if not self.build.scan:
        run_ohs_build()
```

Reply:
> we might be doing a "dirty read"

Where? How? You are in the middle of a transaction, all other workers are waiting to enter their transactions.

Reply:
Oh, my bad. I didn't imagine the transaction as blocking 😅
I was thinking we were using something like https://en.wikipedia.org/wiki/Optimistic_concurrency_control...

In this case I can make the read. Still I don't like it 🙂 Blocking the db for such a long time I don't think is a good idea.

Reply:
I don't think we are overriding the db defaults via SqlAlchemy, so it should be as you said.
So you are right, we are using a level of isolation that does not allow to have inconsistent readings.
I will keep your code 🙂

From https://www.postgresql.org/docs/current/sql-set-transaction.html the defaults is the following.
```
READ COMMITTED
A statement can only see rows committed before it began. This is the default.
```

Reply:
> ```
> READ COMMITTED
> A statement can only see rows committed before it began. This is the default.
> ```

Right, but unfortunately that doesn't prevent race conditions, because it will allow all workers to continue and commit the changes one after the other, if I understand it correctly. No, I thought SQLAlchemy session works as a locking mechanism, but I'm not so sure anymore. Let me do a test.

Reply:
Ok, I tested it and it indeed doesn't work, we would have to introduce explicit locking with something like `LOCK TABLE sidetags IN ACCESS EXCLUSIVE MODE`, which is far from being elegant.
However I found out one can lock an individual record for updating, with [sqlalchemy.orm.Query.with_for_update()](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.with_for_update), I tested it and it indeed makes all workers block until the one updating the record is finished and commits or rollbacks the transaction. I think that's perfect for our use-case. I suppose your context manager could look something like this:

```python
@contextmanager
def add_scan_for_updating(self):
    with sa_session_transaction(commit=True) as session:
        existing_scan = (
            session.query(OSHScanModel)
            .filter_by(copr_build_target_id=self.id)
            .with_for_update()
            .first()
        )
        if existing_scan:
            new_scan = None
        else:
            new_scan = OSHScanModel()
            new_scan.copr_build_target = self
            session.add(new_scan)
        try:
            yield new_scan
        except Exception:
            session.rollback()
            raise
        else:
            if new_scan:
                session.add(new_scan)
```